renderer.go

```diff
@@ -38,64 +47,74 @@ func (r *Renderer) AddOptions(opts ...renderer.Option) {
 	for name, value := range config.Options {
 		r.config.SetOption(name, value)
 	}
-	// TODO handle any config.NodeRenderers set by opts
+
+	// handle any config.NodeRenderers set by opts
+	config.NodeRenderers.Sort()
+	l := len(config.NodeRenderers)
+	for i := l - 1; i >= 0; i-- {
+		v := config.NodeRenderers[i]
+		nr, _ := v.Value.(renderer.NodeRenderer)
+		nr.RegisterFuncs(r)
+	}
+}
+
+func (r *Renderer) Register(kind ast.NodeKind, fun renderer.NodeRendererFunc) {
+	r.nodeRendererFuncsTmp[kind] = fun
+	if int(kind) > r.maxKind {
+		r.maxKind = int(kind)
+	}
 }
 
 // Render implements renderer.Renderer.Render
 func (r *Renderer) Render(w io.Writer, source []byte, n ast.Node) error {
 	r.rc = newRenderContext(w, source, r.config)
-	/* TODO
-	reg.Register(ast.KindString, r.renderString)
-	*/
+	r.initSync.Do(func() {
+		r.nodeRendererFuncs = make([]nodeRenderer, r.maxKind+1)
+		// add default functions
+		// blocks
+		r.nodeRendererFuncs[ast.KindDocument] = r.renderBlockSeparator
+		r.nodeRendererFuncs[ast.KindHeading] = r.chainRenderers(r.renderBlockSeparator, r.renderHeading)
+		r.nodeRendererFuncs[ast.KindBlockquote] = r.chainRenderers(r.renderBlockSeparator, r.renderBlockquote)
+		r.nodeRendererFuncs[ast.KindCodeBlock] = r.chainRenderers(r.renderBlockSeparator, r.renderCodeBlock)
+		r.nodeRendererFuncs[ast.KindFencedCodeBlock] = r.chainRenderers(r.renderBlockSeparator, r.renderFencedCodeBlock)
+		r.nodeRendererFuncs[ast.KindHTMLBlock] = r.chainRenderers(r.renderBlockSeparator, r.renderHTMLBlock)
+		r.nodeRendererFuncs[ast.KindList] = r.chainRenderers(r.renderBlockSeparator, r.renderList)
+		r.nodeRendererFuncs[ast.KindListItem] = r.chainRenderers(r.renderBlockSeparator, r.renderListItem)
+		r.nodeRendererFuncs[ast.KindParagraph] = r.renderBlockSeparator
+		r.nodeRendererFuncs[ast.KindTextBlock] = r.renderBlockSeparator
+		r.nodeRendererFuncs[ast.KindThematicBreak] = r.chainRenderers(r.renderBlockSeparator, r.renderThematicBreak)
+
+		// inlines
+		r.nodeRendererFuncs[ast.KindAutoLink] = r.renderAutoLink
+		r.nodeRendererFuncs[ast.KindCodeSpan] = r.renderCodeSpan
+		r.nodeRendererFuncs[ast.KindEmphasis] = r.renderEmphasis
+		r.nodeRendererFuncs[ast.KindImage] = r.renderImage
+		r.nodeRendererFuncs[ast.KindLink] = r.renderLink
+		r.nodeRendererFuncs[ast.KindRawHTML] = r.renderRawHTML
+		r.nodeRendererFuncs[ast.KindText] = r.renderText
+		// TODO: add KindString
+		// r.nodeRendererFuncs[ast.KindString] = r.renderString
+
+		for kind, fun := range r.nodeRendererFuncsTmp {
+			r.nodeRendererFuncs[kind] = r.transform(fun)
+		}
+		r.nodeRendererFuncsTmp = nil
+	})
 	return ast.Walk(n, func(n ast.Node, entering bool) (ast.WalkStatus, error) {
-		return r.getRenderer(n)(n, entering), r.rc.writer.Err()
+		return r.nodeRendererFuncs[n.Kind()](n, entering), r.rc.writer.Err()
```

Comment:
The only thing I suggest changing is to add a length check here. Since you're keeping `maxKind` up to date in the `Register` method, I don't think an index out of range issue is likely to come up in practice. If it _does_ happen, it'll result in a panic.

The existing version of the code would turn this into a no-op by calling the function returned by `chainRenderers(...)` with an empty slice. We can maintain that behavior here by returning `ast.WalkContinue`.